File: client/src/textSimilarity/tfIdfVectorizer.ts
import { l2Normalize } from "./utils";

export default class TfIdfVectorizer {
    private vocab: Map<string, number>
    private idf: Map<string, number>
    private corpusVectors: number[][]

    public constructor(corpus: string[][]) {
        const df = this.calculateDocumentFrequency(corpus);
        this.idf = this.calculateInverseDocumentFrequency(df, corpus);

        const terms = Array.from(df.keys()).sort();
        this.vocab = new Map();
        terms.forEach((term, i) => this.vocab.set(term, i));

        this.corpusVectors = corpus.map(doc => this.vectorize(doc));
    }

    private calculateTermFrequency(doc: string[]) {
        const tf: Map<string, number> = new Map();

        const length = doc.length;
        if (length <= 0) {
            return tf;
        }

        for (const term of doc) {
            tf.set(term, (tf.get(term) ?? 0) + 1);
        }

        for (const [term, count] of tf) {
            tf.set(term, count / length);
        }

        return tf;
    }

    private calculateDocumentFrequency(docs: string[][]) {
        const df: Map<string, number> = new Map();

        for (const doc of docs) {
            const uniqueTerms = new Set(doc);
            for (const term of uniqueTerms) {
                df.set(term, (df.get(term) ?? 0) + 1);
            }
        }
        return df;
    }

    private calculateInverseDocumentFrequency(df: Map<string, number>, docs: string[][]) {
        const nDocs = docs.length;

        const idf: Map<string, number> = new Map;
        for (const [term, freq] of df) {
            idf.set(term, Math.log((nDocs + 1) / (freq + 1)) + 1);
        }
        return idf;
    }

    public vectorize(tokens: string[]) {
        const vector = new Array(this.vocab.size).fill(0);

        const tf = this.calculateTermFrequency(tokens);

        for (const [term, tfVal] of tf) {
            const idx = this.vocab.get(term);
            if (idx !== undefined) {
                vector[idx] = tfVal * (this.idf.get(term) ?? 0);
            }
        }
        return l2Normalize(vector);
    }

    public getCorpusVectors() {
        return this.corpusVectors;
    }
}

File: server/controllers/encoders/tfidf_vectorizer.py
from collections import Counter
import numpy as np
from controllers.encoders.encoder import Encoder
from typing import Callable
from utils.vector_numpy import l2_normalize
from schemas.similarity import SearchMethod

class TfIdfVectorizer(Encoder):
	def __init__(self, preprocessor_fn: Callable[[str], list[str]], use_idf: bool = True, sublinear_tf: bool = True, smooth_idf: bool = True):
		super().__init__(SearchMethod.TFIDF)
		self.preprocessor = preprocessor_fn
		self.fitted = False
		self.use_idf = use_idf
		self.sublinear_tf = sublinear_tf
		self.smooth_idf = smooth_idf

	def calculate_term_frequency(self, tokens: list[str]):
		tf = np.zeros(len(self.vocab))
		doc_len = len(tokens)

		if doc_len <= 0:
			return tf
		
		counts = Counter(tokens)

		if self.sublinear_tf:
			# TF sublineal: 1 + log(term_frequency)
			for term, count in counts.items():
				idx = self.vocab.get(term)
				if idx is not None:
					tf[idx] = 1 + np.log(count / doc_len)
		else:
			for term, count in counts.items():
				idx = self.vocab.get(term)
				if idx is not None:
					tf[idx] = count / doc_len
		return tf
	
	def _calculate_document_frequency(self, corpus_tokens: list[list[str]]):
		df = Counter()

		for tokens in corpus_tokens:
			unique_term = set(tokens)
			df.update(unique_term)

		return df
	
	def _calculate_inverse_document_frequency(self, df: Counter, n_docs: int):
		idf = np.zeros(len(self.terms))
		for i, term in enumerate(self.terms):
			doc_freq = df.get(term, 0)
			if self.smooth_idf:
				idf[i] = np.log((n_docs + 1) / (doc_freq + 1)) + 1
			else:
				idf[i] = np.log(n_docs / (doc_freq + 1)) + 1
		return idf

	def fit(self, texts: list[str]):
		tokenized_docs = [self.preprocessor(text) for text in texts]

		n_docs = len(tokenized_docs)

		all_terms = set()
		for tokens in tokenized_docs:
			all_terms.update(tokens)
		self.terms = sorted(all_terms)
		self.vocab = {term: i for i, term in enumerate(self.terms)}

		if self.use_idf:
			df = self._calculate_document_frequency(tokenized_docs)
			self.idf = self._calculate_inverse_document_frequency(df, n_docs)
		else:
			self.idf = np.ones(len(self.terms))

		self.fitted = True

		return self
	
	def transform(self, texts: list[str], normalize: bool = True):
		if not self.fitted:
			raise ValueError("Vectorizer not fitted. Call 'fit' first.")
		
		tokenized_docs = [self.preprocessor(text) for text in texts]
		
		X = np.zeros((len(tokenized_docs), len(self.terms)))

		for i, tokens in enumerate(tokenized_docs):
			tf = self.calculate_term_frequency(tokens)
			X[i] = tf * self.idf

		if normalize:
			return l2_normalize(X)
		return X
	
	def get_feature_names(self):
		if not self.fitted:
			raise ValueError("Vectorizer not fitted.")
		return self.terms

	def get_idf(self):
		if not self.fitted:
			raise ValueError("Vectorizer not fitted.")
		if not self.use_idf:
			raise ValueError("IDF not computed because use_idf=False.")
		return self.idf
	
	def get_vocab(self):
		if not self.fitted:
			raise ValueError("Vectorizer not fitted.")
		return self.vocab
